### ISSUE TITLE
[HOLD] Destructure props of all function components by adding an ESLint rule running the linter with --fix

### ISSUE THREAD

**Issue by hayata-suenaga** [Expensify/App#22671]
HOLD on https://github.com/Expensify/App/pull/20179

[Per this discussion](https://github.com/Expensify/App/pull/21478#issuecomment-1626726407) on changing the guideline to always requiring prop destructuring to assign default values to optional props.

Add the following config

```
'react/destructuring-assignment': ['error', 'always', {ignoreClassFields: true, destructureInSignature: 'always'}],
```

and run `--fix`

**Comment by blazejkustra** [comment]
@hayata-suenaga Why not add the rule now in [TS setup PR](https://github.com/Expensify/App/pull/20179#pullrequestreview-1524917442)? If we put it in rules for TS files, we wouldn't have to run eslint --fix at all. When migrating components to TS we could remove defaultProps and use prop destructuring.

**Comment by hayata-suenaga** [comment]
I was thinking doing that but @roryabraham suggested we ran `--fix` before starting the migration

and if we gonna do that, we should probably do that in a separate PR

**Comment by roryabraham** [comment]
I don't feel strongly either way, but figured since the rule supports `--fix` we can get it out of the way now and reduce the diff of TS migration PRs.

**Comment by hayata-suenaga** [comment]
Will do this before the migration will start

**Comment by hayata-suenaga** [comment]
will put PR up tomorrow

**Comment by hayata-suenaga** [comment]
This is on hold for https://github.com/Expensify/App/pull/20179 which will turn off an ESLint setting that conflicts with de-structured props

**Comment by hayata-suenaga** [comment]
now that #20179 has been merged, I'll work on this today

**Comment by hayata-suenaga** [comment]
Turned out that ESLint cannot fix all de-structuring rule violations automatically.

These are only files where ESLint was able to fix code automatically.

<img width="300" alt="Screenshot 2023-08-07 at 9 02 52 AM" src="https://github.com/Expensify/App/assets/98560306/fc4df1c3-63d9-4e6a-8bec-4fdc2023b02f">

I also tried to run ESLint with the latest version (the current version specified in `package.json` is `7.32.0` but the latest is `8.46.0`.

Current version confirmation
```
hayatasuenaga@Hayatas-Work-MBP App % npm list eslint --depth=0
new.expensify@1.3.50-3 /Users/hayatasuenaga/Expensidev/App
└── eslint@7.32.0
```

Running the latest ESLint. Still doesn't fix issues automatically
<img width="600" alt="Screenshot 2023-08-07 at 9 01 03 AM" src="https://github.com/Expensify/App/assets/98560306/7a3a3477-006d-43ba-ac62-387b93b499e5">

@roryabraham what do you think? should we let migration PR also handle prop de-structuring? (we can specify the rule only for `ts` and `tsx` rules so that when engineers migrate a file to a TS file, they gonna be notified about de-structuring rule violation if they didn't de-structure

**Comment by hayata-suenaga** [comment]
updated Rory about the limit of `--fix` option [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1692138509518109?thread_ts=1689968108.162539&cid=C01GTK53T8Q) in slack

**Comment by hayata-suenaga** [comment]
Closing this per this comment: https://expensify.slack.com/archives/C01GTK53T8Q/p1692139684018249?thread_ts=1689968108.162539&cid=C01GTK53T8Q
